src/utils/getEffectiveTLDPlusOne.ts

```diff
@@ -64,14 +68,93 @@ function createTrie(): TrieNode {
 function search(domain: string): string | null {
   let node = trie
 
-  for (let i = 0; i < domain.length; i++) {
+  let i = 0
+  for (i = 0; i < domain.length; i++) {
     if (node.children.has(domain[i])) {
       node = node.children.get(domain[i])!
+    } else if (node.children.has('*')) {
+      node = node.children.get('*')!
     } else {
-      return node.end ? node.suffix : null
+      break
     }
   }
-  return node.end ? node.suffix : null
+
+  // PSL has rules like *.kobe.jp which covers all third-level domains,
+  // but also has a rule !city.kobe.jp which means exclusion from the previous one
+  if (node.excluded) {
+    let currNode = node
+    while (currNode.parent != null) {
+      currNode = currNode.parent
+      if (currNode.end) {
+        return currNode.suffix
+      }
+    }
+    return null
+  }
+
+  // whole domain string processed
+  if (i == domain.length) {
+    if (node.end) {
+      if (!node.excluded) {
+        return node.suffix
+      } else if (node.parent) {
+        return node.parent.suffix
+      }
+      return node.suffix
+    } else {
+      return null
+    }
+  }
+
+  // part of domain string was processed, but some rule found
+  // so need to find previous matching rule
+  // for example, the loop find node `ar.com` for `fdsaar.com`,
+  // so need to proceed with `.com` here
+  let j
+  for (j = i; j > 0; j--) {
+    if (domain[j] == '.') {
+      break
+    }
+  }
+  const sub = reverse(domain.substring(0, j))
+  let currNode = node
+  while (currNode.parent != null) {
+    if (compare(currNode.suffix, sub)) {
+      return currNode.suffix
+    } else {
+      currNode = currNode.parent
+    }
+  }
+
+  if (currNode != null) {
```

Comment:
Can `currNode` ever be `null`? Seems like no

```suggestion
  return currNode.suffix
```